### ISSUE TITLE
XBOX controller stop working on host when starts DUO

### ISSUE THREAD

**Issue by dgrouev** [DuoStream/Duo#111]
Hello, i am using Patreon DUO version 1.4.4. When my XBOX controller is connected to the host and DUO starts then instantly my XBOX controller stops working. It is still showing up as a connected. I`ve tried from Troubleshooting guide but nothings works. Any ideas?

**Comment by Black-Seraph** [comment]
This is most likely linked to the way Duo jails input devices.

There's two parts to it, one that uses a timing-glitch in ViGEmBus to jail virtual devices into the remote sessions they belong into by setting the session ID device attribute right as the device spawns, which works fairly well, and a second piece that that does the same for console session devices (aka. physical devices), which is much less reliable due to the fact that the arrival of said devices can't be predicted.

If I had to explain it in layman's terms...

- Duo knows when a new virtual gamepad device is going to spawn because its preceeded by Moonlight sending at least one gamepad input event which allows us to employ the ViGEmBus timing-glitch in combination with the session ID device attribute to essentially jail the input device into the session it should be in rather than being host-global
- Physical devices on the other hand have no plug-in precursor event Duo could attach to to set its session ID device attribute before the device initializes fully, which means we have no choice but to operate a polling loop and hope for the best, which, often, isn't enough to get the device configured properly in time causing the issue you describe

The proper solution to all of this would be to implement a HID filter driver, but I lack the required access to Microsoft's hardware developer dashboard to double-sign my own drivers, which means I can't actually deploy such a driver to regular users without first forcing them into test-signing mode and letting the driver run unsigned.

But running unsigned drivers isn't an option either, as that will trigger almost all Anti-Cheat systems out there.

**Comment by Black-Seraph** [comment]
@dgrouev I looked into this some more and made a variety of changes to the gamepad jailing code in version 1.4.8 to address parts of this problem.

I did run into another issue while working on this though, see the latest point added to the Known Issues section to read about it.

**Comment by camilofabregas** [comment]
@Black-Seraph sorry for the inconvenience. But is there a way to make the Xbox controller work? Since I have installed Duo, my controller doesn't connect to my PC anymore. I'm using the original Microsoft Wireless Adapter.

**Comment by XeosCeleres** [comment]
I confirm. If you use an Xbox controller with a Windows USB Adapter. The minute a DUO instance starts, the controller will power off.

**Comment by m33ts4k0z** [comment]
Hello. 

Running 1.4.8 Patreon and Windows 24H2 still suffers from this issue **BUT** only when the controller connects via bluetooth. USB seems to be working correctly. I tried with my xbox one v2 BT controller. Only way to fix the bluetooth is to stop the Duo service and reconnect the controller.

This is a huge issue actually and it defeats the purpose of using Duo at the first place.

A solution or suggestion would be much appreciated.

**Comment by camilofabregas** [comment]
Hi @Black-Seraph I hope this issue is addressed in the upcoming version of Duo. 

It's to this day that I haven't been able to use your software because the official Microsoft Wirless Adapter just won't work. The second I start a Duo instance, the controller will instantly disconnect and there's absolutely no way to get it to work while Duo is running.

**Comment by Wacheee** [comment]
This is still an issue, I have an Xbox One, Xbox 360 and a DualSense (PS5) controllers. 

I notice that the only one that has this problem is Xbox One controller, the others doesnt "disconnect" or stop working after I start Duo. So for now my workaround is to use Dualsense/Xbox 360 in host PC and Xbox One in remote session.

The problem is that I want to use DualSense in my remote session because is the one that has the lowest input lag (and in remote sessions its better to have lower input lag).

I tried Sandbox mode and it seems to work, but a lost a lot of savegames, configurations, videogame directories (multiples SSD), so for me its not a solution right now.

Other things that I tried is to use third parties softwares to try to trick windows thinking that Xbox One controller is an Xbox 360 one but it didnt work. If you know another software that can do the trick it would be awesome!

**Comment by Wacheee** [comment]
After trying a lot of free softwares (x360ce with HIDguardian, Inputmapper with HIDhide, XOutput) none of those worked. The only one that work good is reWASD that can change (emulate) my Xbox One Controller to an Xbox 360 Controller, by doing this and then start Duo i can still using this controllers in host PC without problems:
![image](https://github.com/user-attachments/assets/28cce360-81a5-42ee-84cc-80db89a71d46)

The donwside is that is an expensive software (subscription or one expensive payment). 

@Black-Seraph maybe if you can find a way to turn the xbox one controllers from the host machine to xbox 360 one maybe it will solve this issue.

**Comment by MadWizardDE** [comment]
Hello @Wacheee and thank you for sharing your findings! While checking that everything works to my necessities with Duo, I rather forgot the quite obvious use case, trying to use a host connected gamepad while Duo is running – until now. It really seems to have something to do with using a Xbox One Gamepad, though. Apparently reWASD manage to grab the gamepad input on a rather low level, because it doesn't seem to have problems recognizing it while Duo is running, when Steam doesn't even see that there is a gamepad connected. What it seems to do is, disabling the gamepad by itself and creating another virtual Xbox 360 device, that then can be recognized easily.

Until now I didn't test, if that gamepad actually bleeds into the virtual sessions, created by Duo – but then again that is a rather non-usecase for me, because I don't plan to use gamepads in virtual sessions, only in the console session. Maybe this could be a rather simple fix, for some users? At least offering the option, to disable gamepad jailing, for those who doesn't need it anyway, or just do it in virtual sessions. I tried to uninstall the ViGEmBus driver (and keep it uninstalled), but that did not make any difference.

reWASD is quite expensive, but least seems to be a rather powerful tool for that money. It's good to know having a solid workaround. But I definitely think we need this in the main source in not quite a too far future.

**Comment by m33ts4k0z** [comment]
Guys this issue is already resolved with sandbox in 1.4.9 on 23H2 hosts and soon will even for 24H2 with 1.5.0. No need to use any other software.

**Comment by Wacheee** [comment]
@m33ts4k0z Yes, indeed, with Sandbox the problem is solved, but in my case I dont use sandbox because I lost a lot of savegames, configurations, videogame directories. So for now I have to use reWASD.

**Comment by MadWizardDE** [comment]
Yeah I believe you that it's no problem if you use Sandbox, although I find it strange, because for me it happens even without any Duo instance started, just the base service running. But since I have 24H2 installed, I cannot really check for now.

Apart from this, I have to agree with @Wacheee, that sandboxing Duo cannot be the only solution, because virtualization takes up resources and in my case that would be totally unnecissary, because the two user accounts should behave like normal user accounts on the PC without the complications and overhead of a sandbox. Nonetheless I appreciate the work that is being done to make sandboxes a possibility. Just want to make clear, that it shouldn't be a must-do, if you want to use gamepads, locally.

**Comment by Natanji** [comment]
I don't know about Sandbox mode, never used that; but I can confirm that it also affects Xbox Series S/X controllers. My setup is having my PC connected to my TV so the kids can play, and me streaming to my Steam Deck so I can play my more demanding (and Xbox Game Pass) games simultaneously. So the host needs to be able to use physically connected controllers, while Duo should only capture the virtual controllers that Moonlight uses.

I can also confirm that only Xbox controllers are affected. When I connect any of my two Xbox controllers to the Host, they show up in Windows 11 (23H2) under devices, but can't be used inside games. Even when reconnecting them while starting duo. When I stop the Duo service, they need to be reconnected, otherwise they still don't work. But then they do.

My workaround for now is to use two PS5 Dualsense controllers, also connected via Bluetooth to the Host - they aren't affected by the issue, it works flawlessly. Steam input enabled or disabled doesn't make a difference. I'm using current Duo 1.5.2.

I don't know what makes Duo behave differently for Xbox and PS controllers but it would be awesome if this could get fixed!

**Comment by Black-Seraph** [comment]
@Natanji XBOX controllers on Windows are somewhat special in the way that they use a different driver than regular HID (human input devices) do.

While all other gamepads, steering wheels, pedals, flight sticks, etc. all use the regular usbhid.sys driver, XBOX controllers and certain other XBOX-branded peripherals, use the xusb*.sys driver.

The downside to using the xusb*.sys driver is that the driver is marked as non-stoppable, which means the device can't be restarted in a software-controlled way, unlike regular HID devices using the usbhid.sys driver.

To properly jail input devices though, the following set of events must come to pass:

1. The device must be connected and go through its arrival process.
2. The device's session ID property must be set.
3. The device must be stopped and then restarted.

Step 3 is where things often go sour on xusb*.sys based devices, as the driver simply doesn't support stop and start requests.

This step is of course just a best practice, and depending on process start timings, etc. the isolation might work out just fine even without it, but in reality, it often doesn't.

Dualsense, and really any other gamepad that isn't XBOX-branded, doesn't have this issue as we can safely restart the device as part of its isolation process.

tl;dr: This is a driver thing that Duo can't fix, and Duo is already doing everything in its power to properly isolate gamepads.

The only way this process could be made bulletproof is if we wrote our own HID filter driver, but that isn't possible because we're barred from signing said driver with Microsoft, which means we can write the driver, but we have no way to ship it to endusers.
